src/Interpreters/SquashingTransform.cpp

```diff
@@ -126,4 +127,138 @@ bool SquashingTransform::isEnoughSize(size_t rows, size_t bytes) const
         || (min_block_size_bytes && bytes >= min_block_size_bytes);
 }
 
+NewSquashingTransform::NewSquashingTransform(Block header_, size_t min_block_size_rows_, size_t min_block_size_bytes_)
+    : min_block_size_rows(min_block_size_rows_)
+    , min_block_size_bytes(min_block_size_bytes_)
+    , header(std::move(header_))
+{
+}
+
+Block NewSquashingTransform::add(Chunk && input_chunk)
+{
+    return addImpl(std::move(input_chunk));
+}
+
+Block NewSquashingTransform::addImpl(Chunk && input_chunk)
+{
+    if (!input_chunk.hasChunkInfo())
+    {
+        Block to_return;
+        std::swap(to_return, accumulated_block);
+        return to_return;
+    }
+
+    const auto *info = getInfoFromChunk(input_chunk);
+    for (auto & chunk : info->chunks)
+        append(chunk.clone());
+
+    {
+        Block to_return;
+        std::swap(to_return, accumulated_block);
+        return to_return;
+    }
+}
+
+void NewSquashingTransform::append(Chunk && input_chunk)
+{
+    if (input_chunk.getNumColumns() == 0)
+        return;
+    if (!accumulated_block)
+    {
+        for (size_t i = 0; i < input_chunk.getNumColumns(); ++ i)
+        {
+            ColumnWithTypeAndName col = ColumnWithTypeAndName(input_chunk.getColumns()[i], header.getDataTypes()[i], header.getNames()[i]);
+            accumulated_block.insert(accumulated_block.columns(), col);
+        }
+        return;
+    }
+
+    for (size_t i = 0, size = accumulated_block.columns(); i < size; ++i)
+    {
+        const auto source_column = input_chunk.getColumns()[i];
+
+        auto mutable_column = IColumn::mutate(std::move(accumulated_block.getByPosition(i).column));
+        mutable_column->insertRangeFrom(*source_column, 0, source_column->size());
+        accumulated_block.getByPosition(i).column = std::move(mutable_column);
+    }
+}
+
+const ChunksToSquash* NewSquashingTransform::getInfoFromChunk(const Chunk & chunk)
+{
+    const auto& info = chunk.getChunkInfo();
+    const auto * agg_info = typeid_cast<const ChunksToSquash *>(info.get());
+
+    return agg_info;
+}
+
+BalanceTransform::BalanceTransform(Block header_, size_t min_block_size_rows_, size_t min_block_size_bytes_)
+    : min_block_size_rows(min_block_size_rows_)
+    , min_block_size_bytes(min_block_size_bytes_)
+    , header(std::move(header_))
+{
+}
+
+Chunk BalanceTransform::add(Block && input_block)
+{
+    return addImpl(std::move(input_block));
+}
+
+Chunk BalanceTransform::addImpl(Block && input_block)
+{
+    Chunk input_chunk(input_block.getColumns(), input_block.rows());
```

Comment:
Lets avoid converting from chunk to block and again to chunk here. I think the first stage could be done with chunks only.